packages/doenetml-worker-rust/lib-doenetml-core/src/dep_graph/graph.rs

```diff
@@ -0,0 +1,307 @@
+use std::{collections::HashMap, fmt::Debug, hash::Hash};
+
+/// An abstract graph. The speed of lookups is determined by the `Taggable` type,
+/// which is used to map `Node -> usize` for lookups in the internal data structures.
+///
+/// Example using `HashMap` as the `Taggable` type:
+/// ```rust
+/// # use doenetml_core::dep_graph::graph::Graph;
+/// use std::collections::HashMap;
+/// // Make a graph
+/// // a -> b -> c
+/// // a -> c
+/// let mut graph = Graph::<String, HashMap<_, _>>::new();
+/// graph.add_node("a".into());
+/// graph.add_node("b".into());
+/// graph.add_node("c".into());
+/// graph.add_edge("a".into(), "b".into());
+/// graph.add_edge("a".into(), "c".into());
+/// graph.add_edge("b".into(), "c".into());
+/// assert_eq!(graph.walk_descendants("a".into()).collect::<Vec<_>>(), vec!["b", "c"]);
+/// ```
+#[derive(Debug)]
+pub struct Graph<Node: Clone + Eq + Hash + Debug, Index: Taggable<Node, usize>> {
+    /// An `Taggable` that allows for looking up the index of a node in the graph.
+    /// I.e., used for `Node -> usize` lookups.
+    index: Index,
+    /// The nodes in the graph. Internally nodes are represented as `usize`. But `Graph` stores nodes
+    /// for the users to make its internal structure opaque.
+    nodes: Vec<Node>,
+    /// Directed edges between nodes.
+    edges: Vec<Vec<usize>>,
+    /// Reverse directed version of every edge.
+    reverse_edges: Vec<Vec<usize>>,
+}
+
+impl<Node: Clone + Eq + Hash + Debug, Index: Taggable<Node, usize>> Graph<Node, Index> {
+    /// Add a node to the graph.
+    pub fn add_node(&mut self, node: Node) {
+        let index = self.nodes.len();
+        self.index.set_tag(node.clone(), index);
+        // There should always be the same number of elements in the `nodes`, `edges`, and `reverse_edges`
+        // vectors.
+        self.nodes.push(node);
+        self.edges.push(Vec::new());
+        self.reverse_edges.push(Vec::new());
+    }
+    /// Set an edge between two nodes.
+    pub fn add_edge(&mut self, from: Node, to: Node) {
+        let &from_index = self.index.get_tag(from);
```

Comment:
The data structure could handle duplicate edges between nodes. I'm not sure a reason for supporting such a "multigraph", as it would seem easier to reason about the behavior of `Graph` if we allowed only one edge between each pair of nodes. In that case, maybe check for a duplicate and panic if a duplicate edge is created?

Reply:
Especially if we call it a "directed acyclic graph", then we should make sure we don't have multiple edges.

Reply:
I left out a check for efficiency. I could put one in, but the naive implementation could lead to quadratic insertion behavior, and putting a hashset in is probably overkill...

Reply:
As long it doesn't have strange behavior with multiple edges, it's probably not worth doing anything about it.